### PR TITLE
[issue:60] Fix partition topic message router logic

### LINKED ISSUE #60: The flow of multi-Partition consumer can't  work
#### Expected behavior

The flow of multi-Partition consumer can correct  work

#### Actual behavior

The receive blocked.

#### Steps to reproduce

```
func TestPartitionTopicsConsumerPubSub(t *testing.T) {
	client, err := NewClient(ClientOptions{
		URL: lookupURL,
	})
	assert.Nil(t, err)
	defer client.Close()

	topic := "persistent://public/default/testGetPartitions-2"
	testURL := adminURL + "/" + "admin/v2/persistent/public/default/testGetPartitions-2/partitions"

	makeHTTPCall(t, http.MethodPut, testURL, "3")

	// create producer
	producer, err := client.CreateProducer(ProducerOptions{
		Topic: topic,
	})
	assert.Nil(t, err)
	defer producer.Close()

	topics, err := client.TopicPartitions(topic)
	assert.Nil(t, err)
	assert.Equal(t, topic+"-partition-0", topics[0])
	assert.Equal(t, topic+"-partition-1", topics[1])
	assert.Equal(t, topic+"-partition-2", topics[2])

	consumer, err := client.Subscribe(ConsumerOptions{
		Topic:             topic,
		SubscriptionName:  "my-sub-1",
		Type:              Exclusive,
		ReceiverQueueSize: 10,
	})
	assert.Nil(t, err)
	defer consumer.Close()

	ctx := context.Background()
	for i := 0; i < 500; i++ {
		err := producer.Send(ctx, &ProducerMessage{
			Payload: []byte(fmt.Sprintf("hello-%d", i)),
		})
		assert.Nil(t, err)
	}

	msgs := make([]string, 0)

	for i := 0; i < 500; i++ {
		msg, err := consumer.Receive(ctx)
		assert.Nil(t, err)
		msgs = append(msgs, string(msg.Payload()))

		fmt.Printf("Received message msgId: %#v -- content: '%s'\n",
			msg.ID(), string(msg.Payload()))
		//
		//if err := consumer.Ack(msg); err != nil {
		//	assert.Nil(t, err)
		//}
	}

	assert.Equal(t, len(msgs), 500)
}
```

#### System configuration
**Pulsar version**: x.y

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <ranxiaolong716@gmail.com>

Fixes #60

### Motivation

- Fix partition topic message router logic
- Fix permits of flow data race

### CODE REVIEW

**Comment by sijie** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

File: pulsar/impl_partition_consumer.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ import (
 	"fmt"
 	"math"
 	"sync"
+	`sync/atomic`
```
Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: different mark of '`' , maybe use "\"" is better

File: pulsar/impl_partition_consumer.go [hunk 2]
```diff
@@ -265,21 +264,21 @@ func (pc *partitionConsumer) trackMessage(msgID MessageID) error {
...
 	// send flow request after 1/2 of the queue has been consumed
 	if pc.receivedSinceFlow >= highWater {
 		pc.log.Debugf("send flow command to broker, permits size is: %d", pc.receivedSinceFlow)
-		err := pc.internalFlow(pc.receivedSinceFlow)
+		err := pc.internalFlow(highWater)
 		if err != nil {
 			pc.log.Errorf("Send flow cmd error:%s", err.Error())
-			pc.receivedSinceFlow = 0
+			atomic.SwapUint32(&pc.receivedSinceFlow, 0)
+			//pc.receivedSinceFlow = 0
```
Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: remove extra code better

File: pulsar/impl_partition_consumer.go [hunk 3]
```diff
@@ -312,23 +311,53 @@ func (pc *partitionConsumer) Receive(ctx context.Context) (message Message, err
...
-					return err
+		for {
+			select {
+			case tmpMsg, ok := <-pc.subQueue:
+				if ok {
+					msgs <- tmpMsg
+
+					err := pc.messageProcessed(tmpMsg.ID())
+					if err != nil {
+						fmt.Println("hello error....")
```
Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: remove fmt pls

File: pulsar/impl_partition_consumer.go [hunk 4]
```diff
@@ -312,23 +311,53 @@ func (pc *partitionConsumer) Receive(ctx context.Context) (message Message, err
...
+						continue CONSUMER
+						//return err
+					}
+					fmt.Printf("In receiveAsync, big chan size:%d\n", len(msgs))
```
Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: remove fmt pls

File: pulsar/impl_partition_consumer.go [hunk 5]
```diff
@@ -610,12 +639,15 @@ func (pc *partitionConsumer) internalFlow(permits uint32) error {
...
 
 	requestID := pc.client.rpcClient.NewRequestID()
+	pc.log.Debugf("send flow cmd to consumer: [%d], permits size: [%d]", pc.consumerID, permits)
 	_, err := pc.client.rpcClient.RequestOnCnxNoWait(pc.cnx, requestID,
 		pb.BaseCommand_FLOW, &pb.CommandFlow{
 			ConsumerId:     proto.Uint64(pc.consumerID),
 			MessagePermits: proto.Uint32(permits),
 		})
 
+	fmt.Printf("Send flow cmd to broker, consumerID: %d, permits: %d \n", pc.consumerID, permits)
```
Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: remove fmt pls

File: pulsar/impl_partition_consumer.go [hunk 6]
```diff
@@ -650,27 +682,31 @@ func (pc *partitionConsumer) MessageReceived(response *pb.CommandMessage, header
 			Consumer: pc,
 		}
 
+		pc.log.Debugf("receive message form broker, payload is:%s", string(payload))
+
+		fmt.Printf("From broker receive msg: %s , consumerID:{%d}\n", string(payload), pc.consumerID)
```
Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: remove fmt pls

File: pulsar/impl_partition_consumer.go [hunk 7]
```diff
@@ -650,27 +682,31 @@ func (pc *partitionConsumer) MessageReceived(response *pb.CommandMessage, header
...
-			}
-
-			if !dup {
-				pc.redeliverMessages = append(pc.redeliverMessages, newMid)
-			}
-			pc.omu.Unlock()
-			continue
-		default:
-			return fmt.Errorf("consumer message channel on topic %s is full (capacity = %d)", pc.Topic(), cap(pc.options.MessageChannel))
+			fmt.Printf("sub queue size is: %d, consumerID: {%d}\n", len(pc.subQueue), pc.consumerID)
```
Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: remove fmt pls

Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: remove fmt pls

File: pulsar/impl_partition_consumer.go [hunk 8]
```diff
@@ -650,27 +682,31 @@ func (pc *partitionConsumer) MessageReceived(response *pb.CommandMessage, header
...
+			return nil
+		//default:
```
Comment (xujianhai666) [inline pulsar/impl_partition_consumer.go]: remove extra code

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by xujianhai666** [comment]
@wolfstudy resolve conflict. pls

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

File: pulsar/impl_producer.go [hunk 9]
```diff
@@ -45,14 +48,18 @@ func newProducer(client *client, options *ProducerOptions) (*producer, error) {
...
+
 	p := &producer{
 		topic: options.Topic,
 	}
 
 	if options.MessageRouter == nil {
 		internalRouter := internal.NewDefaultRouter(
 			internal.NewSystemClock(),
-			getHashingFunction(options.HashingScheme),
+			getHashingFunction(JavaStringHash),
```
Comment (merlimat) [inline pulsar/impl_producer.go]: If a custom message router is not specified, we should use the default router with the configured hashing scheme.

**Comment by jiazhai** [comment]
@wolfstudy Would you please help track this issue once you have time?

**Pull request closed without merging**